### DISCUSSION TITLE
For an Enterprise environment, is there a method for monitoring all artifacts?

### DISCUSSION THREAD

**Discussion by JordanIKlein** [community/community#73258] [Enterprise] [answered]
### Select Topic Area

Question

### Body

Looking for a possible API or another means of accessing all Artifacts to help monitor our data usage. Thanks in advance!

**Comment by b8zeek** [comment]
Greetings buddy, you definitely can use the GitHub REST API to list and manage artifacts for your repositories. Example of how you can list all artifacts for a repository below (replace `{owner}` and `{repo}`; take a look at GitHub REST API documentation, as well):
```
curl \
  -H "Accept: application/vnd.github.v3+json" \
  https://api.github.com/repos/{owner}/{repo}/actions/artifacts
```

This will return an object containing information about all the artifacts for the specified repository. To manage your data usage, you could write a script that fetches this data and calculates the total size of all artifacts.

P.S. To authenticate yourself, you can do this by generating a personal access token and including it in your requests, again the documentation.

Hope this solves your issue. Take care! 🍺 

🍀

**Reply by JordanIKlein** [reply]
Thanks for the assistance! Is there a way that would help obtain all of the artifacts with a single call rather than going into every repo? Lets say I had over 2,000 repos, that would be a lot of calls.

**Comment by b8zeek** [comment, marked as answer]
No buddy, GitHub API does not provide a direct way to fetch all artifacts from all repositories in a single call. You would need to iterate over each repository and make a separate API call to fetch the artifacts for each repository.

Here's a basic pseudocode outline would look like this:
- get a list of all repositories
- for each repository, get a list of the workflows
- for each workflow, get a list of all runs
- for each run, get a list of all artifacts
- download the artifacts.

Bare in mind the pagination as GitHub wouldn't give everything in a single response. Another thing, some API calls might not be able to use `page` parameter, they are cursor based and you would have to check `Link` header which is returned after fetching the first batch.

Also, be aware of the rate limits, I think GitHub allows up to 5.000 requests per hour for authenticated users.

Hope this answers it, let me know how it goes. 🍺 

🍀
